docs: add best practices and a "manage charms" how-to (#1615)
canonical/operator: docs/howto/manage-charms.md

```{note}
In most cases, you'll want to use `--profile=machine` or `profile=kubernetes`.
If you are charming an application built with a popular framework, check if
charmcraft has a {external+charmcraft:ref}`specific profile <tutorial>` for it.

Avoid the default (`--profile=simple`), which provides a demo charm, rather than
a base for building a charm of your own.
```

````{tip}
If your repository will hold multiple charms, or a charm and source for other
artifacts, like a Rock, create a `charms` folder at the top level, then a folder
for each charm inside of that one, and run `charmcraft --init` in each charm
folder. You'll end up with a structure similar to:

```
my-charm-set-operators/
├── charms
│   ├── my-charm
│   │   ├── charmcraft.yaml
│   │   ├── pyproject.toml
│   │   ├── README.md
│   │   ├── requirements.txt
│   │   ├── src
│   │   │   └── charm.py
│   │   ├── tests
│   │   │   ├── integration
│   │   │   │   └── test_charm.py
│   │   │   └── unit
│   │   │       └── test_charm.py
│   │   └── tox.ini
│   ├── my-charm-dashboard
|   |   └── ...
│   └── my-charm-helper
|   |   └── ...
├── CONTRIBUTING.md
├── LICENSE
├── README.md
└── rock
    └── ...
```

````

> ## Development tooling

Reviewer: Try to find an active title. The rule to have all sections in a how-to active, verb-based, has served us well so far, and helped us write tighter docs, so I wouldn't give it up.

The charmcraft profiles configured recommended tools for developing charms. To
use these, install [tox](https://tox.wiki/en/stable/index.html) on your
development server.

```{tip}

> If you use the `charm-dev` [Multipass](https://canonical.com/multipass) image or

Reviewer: https://canonical.com/multipass/docs/blueprint A blueprint is bigger than an image. 
https://canonical.com/multipass/docs/find-command Lists `charm-dev` in the list of blueprints.

All in all, I still think "blueprint" is the right word here.

It does look like Multipass docs need to be updated so that https://canonical.com/multipass/docs/launch-will say `<image | blueprint>`.

```

- Run `tox` to format and lint the code, and run static type checking and the
  charm unit tests.
- Run `tox -e integration` to run the charm integration tests.
- Run `tox list` to see the available commands.

```{admonition} Best practice
:class: hint

All charms should provide the commands configured by the charmcraft profiles, to
allow easily testing across the charm ecosystem. It's fine to tweak the
configuration of individual tools, or to add additional commands, but keep the
command names and meanings that the profiles provide.
```

## Dependencies

### Python version

Charms run using the Python version provided by the base Ubuntu version. Write
charm code that will run with the Python version of the oldest base you support.

> See also: {external+juju:ref}`Juju | Roadmap and releases <juju-roadmap-and-releases>`

```{admonition} Best practice
:class: hint

Set the [`requires-python`](https://packaging.python.org/en/latest/specifications/pyproject-toml/#requires-python)
version in your `pyproject.toml` so that tooling will detect any use of Python
features not available in the versions you support.
```

### Python packages

> Specify all the all direct dependencies of your charm in your `pyproject.toml`

Reviewer: ```suggestion
Specify all the direct dependencies of your charm in your `pyproject.toml`
```

```toml
# Required group: these are all dependencies required to run the charm.
dependencies = [
    "ops~=2.19",
]

# Required group: these are all dependencies required to run all the charm's tests.
[dependency-groups]
test = [
    "ops[testing]",
    "pytest",
    "coverage[toml]",
    "jubilant",
]
# Optional additional groups:
docs = [
    "canonical-sphinx-extensions",
    "furo",
    "sphinx ~= 8.0.0",
    "sphinxext-opengraph",
]
```

Use the `pyproject.toml` dependencies to specify *all* dependencies (including
indirect or transitive dependencies) in a lock file.

````{admonition} Best practice
:class: hint

When using the `charm` plugin with charmcraft, ensure that you set strict
dependencies to true, for example:

```yaml
parts:
  my-charm:
    plugin: charm
    charm-strict-dependencies: false
```
````

The default lock file is a plain `requirements.txt` file (you can use a tool
such as [pip-compile](https://pip-tools.readthedocs.io/en/latest/) to produce
it from `pyproject.toml`).

```{tip}
Charmcraft provides plugins for {external+charmcraft:ref}`uv <craft_parts_uv_plugin>`
and {external+charmcraft:ref}`poetry <craft_parts_poetry_plugin>`. Use one of

> these tools to simplify generating your lock file.

Reviewer: ```suggestion
these tools to simplify the generation of your lock file.
```

I will defer to your native speaker ear, however.

```

```{admonition} Best practice
:class: hint

Ensure that the `pyproject.toml` *and* the lock file are committed to version
control, so that exact versions of charms can be reproduced.
```

```{tip}
Including an external dependency is a significant choice. It can help with
reducing the complexity and development cost. However, it also increases the
complexity of understanding the entire system, and adds a maintenance burden of
keeping track of upstream versions, particularly around security issues.

> See more: [Our Software Dependency Problem](https://research.swtch.com/deps)
```

> ## Continuous integration

Reviewer: Try to find an active title.

(manage-charms)=
# How to manage charms

> See first: {external+juju:ref}`Juju | Build a charm <build-a-charm>`, {external+charmcraft:ref}`Charmcraft | Manage charms <manage-charms>`

## Create a repository and initialise it

Create a repository with your source control of choice. Commit the code you have
added and changed after every significant change, so that you have a record of
the work, and can revert to an earlier version when required.

```{admonition} Best practice
:class: hint

Name the repository using the pattern ``<charm name>-operator`` for a single
charm, or ``<base charm name>-operators`` when the repository will hold
multiple related charms. For the charm name, see

> {external+charmcraft:ref}`Charmcraft | Specify a name <specify-a-name>`.

Reviewer 2: Do we not recommend what I think is the current pattern of X-operator for the machine charm, and X-k8s-operator for the kubernetes version?

Author: The K8s charm should have a "charm name" of "x-k8s" (except in limited cases), so the repo is then "x-k8s-operator", still following the "charm-operator" scheme.

Reviewer 2: Ah of course, that make sense, thanks.

charmcraft has a {external+charmcraft:ref}`specific profile <tutorial>` for it.

Avoid the default (`--profile=simple`), which provides a demo charm, rather than
a base for building a charm of your own.
```

````{tip}
If your repository will hold multiple charms, or a charm and source for other
artifacts, like a Rock, create a `charms` folder at the top level, then a folder
for each charm inside of that one, and run `charmcraft --init` in each charm
folder. You'll end up with a structure similar to:

```
my-charm-set-operators/
├── charms
│   ├── my-charm
│   │   ├── charmcraft.yaml
│   │   ├── pyproject.toml
│   │   ├── README.md
│   │   ├── requirements.txt
│   │   ├── src
│   │   │   └── charm.py
│   │   ├── tests
│   │   │   ├── integration
│   │   │   │   └── test_charm.py
│   │   │   └── unit
│   │   │       └── test_charm.py
│   │   └── tox.ini
│   ├── my-charm-dashboard
|   |   └── ...
│   └── my-charm-helper
|   |   └── ...
├── CONTRIBUTING.md
├── LICENSE
├── README.md
└── rock
    └── ...
```

````

## Use the provided tooling to maintain style and detect issues early

The Charmcraft profile you've chosen has configured a number of recommended

> tools for developing charms. To use these, install
> [`tox`](https://tox.wiki/en/stable/index.html) on your development server.

Reviewer 2: I wonder if it's worth linking to the `tox` install instructions here to try and create a smoother onboarding experience to using the tools we want charmers to use? The instructions begin by recommending `uv` and `pipx` which is nice. Probably fine as is though, feel free to reject this suggestion.
```suggestion
tools for developing charms. To use these,
[install `tox`](https://tox.wiki/en/stable/installation.html#as-tool)
on your development server.
```

Author: This bit has moved to the charmcraft PR, but did it there.

Hit the ground running with Ops by setting up standard project structure and tools.

> See more: {external+charmcraft:ref}`Charmcraft | Manage Charmcraft <manage-charmcraft>`

Use Charmcraft to quickly initialise your charm project. This generates the
folder structure, creates placeholder configuration and code files, and
configures development tooling.

> See more:
>
> * {external+charmcraft:ref}`Charmcraft | Manage charms > Initialize a charm <initialise-a-charm>` (see also the best practice note on setting up a repository and considering your CI)
> * [Charmcraft | Manage charms > Add charm project metadata, an icon, docs](https://canonical-charmcraft.readthedocs-hosted.com/en/latest/howto/manage-charms/#add-charm-project-metadata-an-icon-docs)

<!--
TODO: Add a reference link in charmcraft for the link above and the 'runtime details' one below, and switch over to external refs.
-->

The essence of a charm is the ``src/charm.py`` file. This is the entry point for
your code whenever Juju emits an event, and defines the interface between Juju

> and the charm workflow.

Author: @dwilding I think this is where you'd add an initial bit about `src/{workload}.py`. We can do that in this PR if you like, or you could do it in a follow-up one later.

I think the main instructions for separating out the code go into the write and structure how-to, but a small mention here would be good.

> See more: {ref}`run-workloads-with-a-charm-kubernetes`, {ref}`run-workloads-with-a-charm-machines`, {ref}`write-and-structure-charm-code`, {ref}`write-unit-tests-for-a-charm`, {ref}`write-integration-tests-for-a-charm`, {ref}`manage-logs`

> These files are, of course, quite generic, so the next thing to do is add
> functionality to your charm. Part of that is also packing, test deploying,
> debugging, and, when ready, publishing your charm on Charmhub.

Reviewer 3: Suggest rephrasing for readability:

```suggestion
The next thing to do is add functionality to your charm.
Then you'll pack, test, and debug your charm.
Finally, when you're ready, you'll publish your charm on Charmhub.
```

```{admonition} Best practice
:class: hint
One of the powers of charms is their reusability. As such, do not try to
duplicate functionality already achieved by an existing charm – rather, make
your charm take advantage of the [charm ecosystem](https://charmhub.io) by
supporting integrating with existing charm solutions for observability,
identity, scaling, and so on.

This also helps you stay compliant with another fundamental rule in charms,
namely that, following the Unix philosophy, each charm should do one thing and
do it well.
```

> 

Author: Something I think would be good here would be a small bit on events: ops lets you observe Juju events (link to details), charm libraries produce custom events that you observe (link to details), ops has its own events (explain or link to lifecycle events).

However, maybe that ought to be in a follow-up PR? I'd like to get this one over the line and it feels like that might be a bigger bit of work. Thoughts?

Reviewer 3: I agree. I think the part above "The next thing to do is add functionality to your charm" naturally leads into a discussion about events and observers. I also agree that a follow-up PR is best.

> See more:
> 
> * Make use of core Juju functionality
>   - {ref}`manage-storage`
>   - {ref}`manage-resources`
>   - {ref}`manage-secrets`
> * Add functionality
>   - [Charmcraft | Add runtime details to a charm](https://canonical-charmcraft.readthedocs-hosted.com/en/latest/howto/manage-charms/#add-runtime-details-to-a-charm)
>   - {ref}`manage-actions`
>   - {ref}`manage-configurations`
>   - {ref}`manage-opened-ports`
> * {external+charmcraft:ref}`Charmcraft | Manage charms > Pack a charm <pack-a-charm>`

> > * {external+juju:ref}`Juju | Manage charms > Deploy a charm <deploy-a-charm>` (you'll need to follow the Deploy a local charm example)

Reviewer 3: Suggest adding quotes for clarity:

```suggestion
> * {external+juju:ref}`Juju | Manage charms > Deploy a charm <deploy-a-charm>` (you'll need to follow the "Deploy a local charm" example)
```

> * {external+juju:ref}`Juju | Manage charms > Debug a charm <debug-a-charm>`
> * {external+charmcraft:ref}`Charmcraft | Publish a charm on Charmhub <publish-a-charm>` (see especially the note on requesting formal review for your charm)

A charm is software: while there can be milestones, there is never a finish
line. So, keep investing in every bit of your charm so that it looks and feels
professional – from polishing metadata (including an icon, a website, docs, and
so on) through polishing features (for example, working to ensure correct and
reliable behavior, adding libraries so people can quickly integrate with your
charm, and so on) all the way to turning it into a successful open source

> project with a community that enjoys it and wants to contribute to it. 

Reviewer 3: :heavy_plus_sign: :heavy_plus_sign: on this sentiment!
